airflow/utils/python_virtualenv.py

```diff
@@ -35,12 +37,35 @@ def _generate_virtualenv_cmd(tmp_dir: str, python_bin: str, system_site_packages
     return cmd
 
 
-def _generate_pip_install_cmd(tmp_dir: str, requirements: List[str]) -> Optional[List[str]]:
+def _generate_pip_install_cmd(tmp_dir: str,
+                              requirements: List[str],
+                              connection_id: Optional[str] = None
+                              ) -> Optional[List[str]]:
     if not requirements:
         return None
-    # direct path alleviates need to activate
-    cmd = [f'{tmp_dir}/bin/pip', 'install']
-    return cmd + requirements
+
+    if connection_id:
+        con: Connection = BaseHook.get_connection(connection_id)
+        user = con.login
+        schema = con.schema or 'https'
+        password = con.get_password()
+        port = con.port or 8080
+        host = con.host
+        host_suffix = con.extra_dejson.get('host_suffix', 'repository/python/simple')
+        if user:
+            index_url = os.path.join(f"{schema}://{user}:{password}@{host}:{port}", host_suffix)
+        else:
+            index_url = os.path.join(f"{schema}://{host}:{port}", host_suffix)
+        private_cmd = [f'{tmp_dir}/bin/pip',
+                       'install',
+                       f'--index-url', index_url,
+                       f'--extra-index-url', 'https://pypi.org/simple'
```

Comment:
This is highly discouraged for very complicated reasons (see the pip issue linked in my previous comment). I recommend setting *only* `index_url` to the custom connection. If you really really must use `--extra-index-url` (you still shouldn’t), set the custom repository as extra and keep the default to PyPI.